### ISSUE TITLE
Problems with repr() in IPython

### ISSUE THREAD

**Issue by ian-r-rose** [sympy/sympy#10062]
Hi all,

This is a bit marginal as to whether it's a SymPy issue or and IPython issue. I've been trying to get some objects to output with LaTeX rendering in IPython notebooks, and I have been getting some sort of  collision with the state of the sympy printing system (regardless of whether my objects actually use sympy). That is to say, if I call `sympy.init_printing()`, it can break the rendering of other objects. In general, is this to be expected? 

To be more concrete, here is a minimal example: Consider a `Parameter` class which knows how to render to LaTex using the IPython function `_repr_latex_()`:
```
class Parameter(object):
    def __init__(self, symbol):
        self.symbol = symbol
    def _repr_latex_(self):
        return '$'+self.symbol+'$'
```
If I create some instances of this, e.g.
```
density = Parameter('\\rho')
gravity = Parameter('g')
```
they output as expected. If I make a list of these, they do not render, as expected, since it is a list:
```
[gravity, density]
```
No matter, I can try to subclass the builtin `list` with a `_repr_latex_()` function:
```
class ParameterList(list):
    def _repr_latex_(self):
        string = '$'
        for n in range(len(self)):
            string += self.__getitem__(n).symbol
            if n != len(self)-1:
                string += ', '
        string += '$'
        return string
```
and this allows me to do what I want with interactive IPython development:
```
plist = ParameterList([density,gravity])
plist
```
which nicely renders the list.

My  problem is, when `sympy.init_printing()` is called, this code breaks. Instead of printing a list of LaTeX typeset strings, it seems to fall back on the default python list `__repr__()`. Browsing through the sympy codebase, it seems that the printing functionality is pretty complicated. Is there reason to expect that it would break the behavior of non-sympy-related code? Any suggestions for how to fix this?

**Comment by asmeurer** [comment]
It looks like a bug in IPython.

SymPy registers display formatters for various builtin types, including `list`. The formatter for `list` only prints LaTeX if all the elements are LaTeX printable (i.e., SymPy types). Otherwise lists that are completely unrelated to SymPy like `[lambda x: x]` would print as LaTeX, and it wouldn't look good. The relevant code is in sympy.interactive.printing.

Since your class is a subclass of list, it is applying to it. This is I guess where the bug is in IPython, since I would expect subclasses to take precedence (even though the `list` printer is registered instead of enabled with a `_repr_` method).  So I would open an issue with them.

In the meantime, two possible workarounds:

- Make your `Parameter` object a SymPy type (subclass from `Basic` or `Expr`). 
- Don't subclass from `list`, but rather `collections.abc.MutableSequence` (I'm not 100% sure this will fix it, but you generally should use `collections.abc` rather than subclassing from built-in containers anyway).

**Comment by moorepants** [comment]
Yes this is a possibility. SymPy has a very extensive printing system that came along before IPython's system. Our current method of enabling out printers to work the the IPython system relies on enabling various classes through a global hook into IPython via `init_printing`. In the long run this interaction with IPython may need to be redesigned.

Our printing system prints lists and tuples of SymPy objects and some other builtins. We may need to rework that. This line may cause the issue:

https://github.com/sympy/sympy/blob/master/sympy/interactive/printing.py#L108

Here is a notebook with the issue: https://gist.github.com/moorepants/5eb4863f65cbe7635dc0

http://nbviewer.ipython.org/gist/moorepants/5eb4863f65cbe7635dc0

**Comment by moorepants** [comment]
Sorry, I had the above half typed yesterday and missed Aaron's response before pressing "Comment".

**Comment by ian-r-rose** [comment]
Thank you for the quick replies.

After some more experimentation, it seems that explicitly registering the output function via the IPython `for_type()` interface as described [here](https://ipython.org/ipython-doc/3/config/integrating.html) does the trick. That is to say, instead of implementing a `_repr_latex_()` function, I can get a handle to the LaTeX formatter and do

```
get_ipython().display_formatter.formatters['text/latex'].\
    for_type(ParameterList, ParameterList.latex_print)
```

I am not sure why this should work while the other approach does not, but this is a fine workaround for my purposes. 

Feel free to close or keep open this issue as you see fit.

**Comment by asmeurer** [comment]
Ah yes, that should also fix it. I would still report this to IPython, though, since it does appear to be a bug in my view.

**Comment by moorepants** [comment]
I tried this with sympy master and IPython 4.0 and it seems to work properly.

**Comment by moorepants** [comment]
Here is the gist showing this is working :http://nbviewer.ipython.org/gist/moorepants/5eb4863f65cbe7635dc0

**Comment by ian-r-rose** [comment]
Huh, neat. I will try to reproduce (I am currently using sympy 0.7.6 and IPython 3.2), but if this is no longer an issue in more recent versions, I see no reason to continue.

**Comment by ian-r-rose** [comment]
Hm, when I try to use the sympy dev version and IPython 4.0 it still does not work.
The workaround described above still fixes it, however.
https://gist.github.com/ian-r-rose/f296376d0d824ee34dbe

**Comment by moorepants** [comment]
I can't reproduce that effect anymore either. I'm not sure what was going on.

**Comment by eric-wieser** [comment]
Can this be closed, based on @moorepants' comments?

**Comment by eric-wieser** [comment]
Just checked out the notebooks - this is still a problem, so cannot be closed.
